### PR TITLE
perf: Fix various N+1 issues on Course API

### PR DESCRIPTION
**Pull request by DawoudSheraz**
### Description

This PR fixes various N+1 issues with Course API and its associated serializers. N+1 issues have been identified with Django Debug toolbar and https://github.com/jmcarp/nplusone. 

- select & prefetch related is missing for some fields when CourseWithProgramSerializer inherits from CourseSerializer. CourseSerializer had those calls in place in [prefetch_queryset](https://github.com/openedx/course-discovery/blob/28ec39978a59223a44ad9ac3046ea730cfb5b7db/course_discovery/apps/api/serializers.py#L1303) method but CourseWithProgramSerializer did not use them and had its own [repeated prefetch](https://github.com/openedx/course-discovery/blob/28ec39978a59223a44ad9ac3046ea730cfb5b7db/course_discovery/apps/api/serializers.py#L1517). 
- Fix N+1 on Related Fields in Serializers (See the note on top on https://www.django-rest-framework.org/api-guide/relations/, also https://www.django-rest-framework.org/api-guide/fields/#source)

For the tests whose query count bumped a bit after fixups:
- CourseViewSetTests::test_options, because of N+1, some extra items are fetched when executing OPTIONS on course endpoint instead of doing it later on demand i.e. lazy. fetch
- TestProgramViewSet::test_retrieve_without_course_runs --This is impacted due to N+1 changes in CourseSerializer on subjects. 

### Testing
- Checkout master discovery on your local
- Hit courses API (api/v1/courses) for listing and fetching a specific course. Do both the actions with ?editable=1 flag, to fetch draft versions.  Keep note of query count from Debug toolbar
- Checkout this branch and repeat above steps. Notice the decrease in query count

Following are count results from my devstack

Action | SQL count on master | SQL count after N+1 fixes
-- | -- | --
Course List (/api/v1/courses) | 344 | 174
Course List - Draft(/api/v1/courses?editable=1) | 370 | 204
Specific course (/api/v1/course/uuid) | 173 | 99
Specific course - Draft (/api/v1/course/uuid?editable=1) | 178 | 109
non-existent UUID | 77 | 16

### CODE REVIEW

File: course_discovery/apps/api/serializers.py [hunk 1]
```diff
@@ -1321,12 +1323,13 @@ def prefetch_queryset(cls, partner, queryset=None, course_runs=None):  # lint-am
...
         ).prefetch_related(
             'expected_learning_items',
+            'level_type__translations',
             'prerequisites',
-            'subjects',
             'collaborators',
             'topics',
             'url_slug_history',
             'url_redirects',
+            'editors',
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/api/serializers.py:1332]: why are these changes made? I'm having a hard time understanding on adding `level_type__translations`, `editors` and removing `subjects` here.

**Review by Ali-D-Akbar — COMMENTED** [review]
Testing locally.

File: course_discovery/apps/api/serializers.py (hunk 1 above)
Comment (DawoudSheraz) [reply course_discovery/apps/api/serializers.py:1332]: See https://docs.djangoproject.com/en/4.2/ref/models/querysets/#select-related & https://docs.djangoproject.com/en/4.2/ref/models/querysets/#prefetch-related

**Review by Ali-D-Akbar — APPROVED** [review]

**Review by AfaqShuaib09 — COMMENTED** [review]
On testing locally, i get following query counts on following endpoints

Endpoint | Master Branch | PR Branch
-- | -- | --
Course List (/api/v1/courses) | 79 | 65
Course List - Draft(/api/v1/courses?editable=1) | 289 | 184
Specific course (/api/v1/course/uuid) | 104 | 78
Specific course - Draft (/api/v1/course/uuid?editable=1) | 123 | 92

File: course_discovery/apps/api/serializers.py [hunk 2]
```diff
@@ -1020,7 +1022,7 @@ class CourseRunSerializer(MinimalCourseRunSerializer):
     def prefetch_queryset(cls, queryset=None):
         queryset = super().prefetch_queryset(queryset=queryset)
 
-        return queryset.select_related('language', 'video').prefetch_related(
+        return queryset.select_related('language', 'video', 'expected_program_type').prefetch_related(
```
Comment (AfaqShuaib09) [inline course_discovery/apps/api/serializers.py:1025]: why we are adding `expected_program_type` here?

Comment (DawoudSheraz) [reply course_discovery/apps/api/serializers.py:1025]: It is FK relation to ProgramType.

**Review by AfaqShuaib09 — APPROVED** [review]

**Pull request merged**